### PR TITLE
feat(routesv2) : Build inbound routes for traffic split apex services

### LINKED ISSUE #2545: Traffic split does not work with routes-v2, no virtual_host for apex service in the inbound route at destination
**Bug description**:
Traffic split does not work with routesV2 when run with the default demo `./demo/run-osm-demo.sh --enable-routes-v2-experimental=true`.

There needs to be a virtual_host corresponding to the apex service in the inbound route configuration at the destination service, but this is not being programmed as expected.

**Affected area (please mark with X where applicable)**:
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [X]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- CI System                  [ ]

**Expected behavior**:
Traffic split should work with routesV2.

**Steps to reproduce the bug (as precisely as possible)**:
1. Apply the following diff for spit policy to be processed correctly
```
--- a/demo/deploy-traffic-split.sh
+++ b/demo/deploy-traffic-split.sh
@@ -12,7 +12,7 @@ metadata:
   name: bookstore-split
   namespace: "$BOOKSTORE_NAMESPACE"
 spec:
-  service: "bookstore.$BOOKSTORE_NAMESPACE"
+  service: bookstore
   backends:
 
   - service: bookstore-v1
```

2. Run the demo:
./demo/run-osm-demo.sh --enable-routes-v2-experimental=true`

**Environment**:
- OSM version (use `osm version`):
`Version: dev; Commit: 1323707da19e7b8590dd06aaf91832c01a4ebb4d; Date: 2021-02-17-10:59`

- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3", GitCommit:"1e11e4a2108024935ecfcb2912226cedeafd99df", GitTreeState:"clean", BuildDate:"2020-10-14T12:50:19Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

- Size of cluster (number of worker nodes in the cluster): 1 node

### PR DESCRIPTION
**Pull request by snehachhabria**
**Description**:

This PR build inbound routes with virtual hosts pointing to the apex
service a for an upstream service that is referenced in a traffic split.
Corresponding unit tests have been added.

I have tested these changes locally by enabling routes v2 on the traffic split e2e tests and they pass for all  traffic split scenarios

Fixes #2545

Signed-off-by: Sneha Chhabria <snchh@microsoft.com>

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution? `no`

### CODE REVIEW

File: pkg/catalog/routes.go [hunk 1]
```diff
@@ -538,13 +586,6 @@ func (mc *MeshCatalog) getDestinationServicesFromTrafficTarget(t *access.Traffic
...
 	inboundPolicies := []*trafficpolicy.InboundTrafficPolicy{}
 
-	// fetch services running workloads with destination service account
-	destServices, err := mc.getDestinationServicesFromTrafficTarget(t)
-	if err != nil {
-		log.Error().Err(err).Msgf("Error resolving destination from traffic target %s (%s)", t.Name, t.Namespace)
-		return inboundPolicies
-	}
-
 	// fetch all routes referenced in traffic target
```
Comment (snehachhabria) [inline pkg/catalog/routes.go:589]: destination services has been removed as discussed in a previous PR : https://github.com/openservicemesh/osm/pull/2593#discussion_r581304432

File: pkg/catalog/routes.go [hunk 2]
```diff
@@ -63,21 +82,50 @@ func (mc *MeshCatalog) listOutboundTrafficPoliciesForTrafficSplits(sourceNamespa
 	return outboundPoliciesFromSplits
 }
 
-// ListOutboundTrafficPolicies returns all outbound traffic policies related to the given service account
-func (mc *MeshCatalog) ListOutboundTrafficPolicies(sa service.K8sServiceAccount) []*trafficpolicy.OutboundTrafficPolicy {
-	outbound := mc.listOutboundPoliciesForTrafficTargets(sa)
+// listPoliciesForTrafficSplits loops through all SMI Traffic Targets resources and returns inbound traffic policies
+// based on when the given service account matches a destination Traffic Target resource having a service referenced in a traffic split
```
Comment (shashankram) [inline pkg/catalog/routes.go:86]: Would you able to break up the function description as follows:
```
// listPoliciesForTrafficSplits loops through all SMI TrafficTarget resources and returns inbound policies for apex services based on the following conditions:
1. the given upstream identity matches the destination specified in a TrafficTarget resource
2. the given list of upstream services are backends specifed in a TrafficSplit resource
```

File: pkg/catalog/routes.go [hunk 3]
```diff
@@ -63,21 +82,50 @@ func (mc *MeshCatalog) listOutboundTrafficPoliciesForTrafficSplits(sourceNamespa
...
-	outboundPoliciesFromSplits := mc.listOutboundTrafficPoliciesForTrafficSplits(sa.Namespace)
-	outbound = trafficpolicy.MergeOutboundPolicies(outbound, outboundPoliciesFromSplits...)
+	for _, t := range mc.meshSpec.ListTrafficTargets() { // loop through all traffic targets
+		if !isValidTrafficTarget(t) {
+			continue
+		}
 
-	return outbound
-}
+		if t.Spec.Destination.Name == upstreamIdentity.Name { // found inbound
```
Comment (shashankram) [inline pkg/catalog/routes.go:95]: Please invert the condition and use `continue` to avoid unnecessary nesting.

File: pkg/catalog/routes.go [hunk 4]
```diff
@@ -63,21 +82,50 @@ func (mc *MeshCatalog) listOutboundTrafficPoliciesForTrafficSplits(sourceNamespa
...
+				log.Error().Err(err).Msgf("Error finding route matches from TrafficTarget %s in namespace %s", t.Name, t.Namespace)
+				continue
+			}
 
-// ListInboundTrafficPolicies returns all inbound traffic policies related to the given service account and inbound services
-func (mc *MeshCatalog) ListInboundTrafficPolicies(sa service.K8sServiceAccount, inboundSvcs []service.MeshService) []*trafficpolicy.InboundTrafficPolicy {
-	inbound := mc.listInboundPoliciesFromTrafficTargets(sa, inboundSvcs)
+			for _, upstreamSvc := range upstreamServices {
+				//check if any of the destination services belong to a traffic split
+				if mc.isTrafficSplitBackendService(upstreamSvc) {
```
Comment (shashankram) [inline pkg/catalog/routes.go:105]: Invert the condition and use `continue`

File: pkg/catalog/service.go [hunk 5]
```diff
@@ -4,47 +4,51 @@ import (
...
-			Namespace: split.ObjectMeta.Namespace,
-		}
-		if reflect.DeepEqual(svc, apexSvc) {
-			return true
+		for _, backend := range split.Spec.Backends {
+			backendService := service.MeshService{
+				Name:      backend.Service,
+				Namespace: split.ObjectMeta.Namespace,
+			}
+			if reflect.DeepEqual(svc, backendService) {
```
Comment (shashankram) [inline pkg/catalog/service.go:22]:
```suggestion
			if svc.Equals(backendService) {
```

**Review by shashankram — APPROVED** [review]

**Review by eduser25 — APPROVED** [review]

**Review by michelleN — COMMENTED** [review]
🎉

**Pull request merged**